<?xml version="1.0"?>
 <!--

      Spawns a rfid equipped thorvald in ncfm facilities.
-->

<launch>
  <!-- PARAMS .............................................................. -->
  <arg name="two_readers"         default="true"/>
  <arg name="w_info_gain"  	      default="0.00"/>
  <arg name="w_travel_distance"  	default="0.2"/>
  <arg name="w_sensing_time"  	  default="0.1"/>
  <arg name="w_battery_status"  	default="0.2"/>
  <arg name="w_rfid_gain"  	      default="0.5"/>
  <arg name="fov"                 default="180"/>
  <arg name="radius"              default="5"/>
  <arg name="coverage_percentage" default="0.95"/>
  <arg name="frontier_threshold"  default="0.001"/>
  <arg name="map_resolution"      default="0.25"/>
  <arg name="dest_folder"  	      default="/home/pulver/Desktop/topoNBS/nbs/"/>
  <arg name="scenario_name"  	    default="INB"/>
  <arg name="use_mcdm"            default="1"/>
  <arg name="num_tags"            default="1"/>
  <arg name="max_iterations"      default="20"/>
  <arg name="distances_map"       default="riseholme_map.csv"/>
  <arg name="pure_args"  	        default="$(arg fov) 
                                           $(arg radius) 
                                           $(arg coverage_percentage) 
                                           $(arg frontier_threshold) 
                                           $(arg map_resolution) 
                                           $(arg w_info_gain) 
                                           $(arg w_travel_distance) 
                                           $(arg w_sensing_time) 
                                           $(arg w_battery_status) 
                                           $(arg w_rfid_gain)
                                           $(arg use_mcdm) 
                                           $(arg num_tags)
                                           $(arg max_iterations)
                                           $(arg distances_map)
                                           $(arg dest_folder)"/>

  <arg name="bag_descriptor" default="$(arg scenario_name)_$(arg w_info_gain)_$(arg w_travel_distance)_$(arg w_sensing_time)_$(arg w_battery_status)_$(arg w_rfid_gain)_date_"/>
  <!-- I highly recomend you that the bag descriptor has correct pure nav params -->
  <arg name="save_bag"  	default="false"/>

  <!-- Topics to record:
    These two were removed because they suppose lots of data
      /nav_grid_visualization_debug/visited
      /plan_grid_visualization_debug/visited
  -->
  <arg name="topic_list" default="/tf /tf_static /mcdm_stats /scan_filtered /map /loc_map /move_base/current_goal /lastTag"/>

  <env name="ROSCONSOLE_CONFIG_FILE"
    value="$(find next_best_sense)/config/custom_rosconsole.conf"/>

  <!-- Launch the grid map visualizer: can be removed after debug!-->
  <node pkg="grid_map_visualization" type="grid_map_visualization" name="nav_grid_visualization_debug" output="screen">
      <rosparam command="load" file="$(find next_best_sense)/config/visualization.yaml" />
  </node>

  <!-- Launch the grid map visualizer: can be removed after debug!-->
  <node pkg="grid_map_visualization" type="grid_map_visualization" name="plan_grid_visualization_debug" output="screen">
      <rosparam command="load" file="$(find next_best_sense)/config/visualization2.yaml" />
  </node>

  <!-- base bringup -->
  <!-- GDB: launch-prefix=gdb -ex run -/-args -->
  <node pkg="next_best_sense" type="pure_navigation" name="nbs" output="screen"  args="$(arg pure_args)" unless="$(arg two_readers)">
    <param name="move_base_costmap_topic_name"          value="/map"/>
    <param name="move_base_costmap_updates_topic_name"  value="/move_base/global_costmap/costmap_updates"/>
    <param name="belief_map_srv_name"                   value="/thorvald/rfid_grid_map_node/get_rfid_belief"/>
    <param name="fake_belief_map_srv_name"              value="/thorvald/rfid_grid_map_node/get_rfid_fake_belief"/>
  </node>
  <node pkg="next_best_sense" type="pure_navigation_two_readers" name="nbs" output="screen"  args="$(arg pure_args)" if="$(arg two_readers)">
    <param name="move_base_costmap_topic_name"          value="/map"/>
    <param name="move_base_costmap_updates_topic_name"  value="/move_base/global_costmap/costmap_updates"/>
    <param name="belief_map_srv_name_left"              value="/thorvald_left/rfid_grid_map_node/get_rfid_belief_left"/>
    <param name="belief_map_srv_name_right"             value="/thorvald_right/rfid_grid_map_node/get_rfid_belief_right"/>
    <param name="fake_belief_map_srv_name_left"         value="/thorvald_left/rfid_grid_map_node/get_rfid_fake_belief_left"/>
    <param name="fake_belief_map_srv_name_right"        value="/thorvald_right/rfid_grid_map_node/get_rfid_fake_belief_right"/>
  </node>

  <!-- start recording the experiment -->
  <group  if="$(arg save_bag)">
    <node pkg="rosbag" type="record" name="bag_recorder"
    args="record  -o $(arg bag_descriptor) $(arg topic_list)" output="screen"/>
  </group>

</launch>
